discojs/src/dataset/dataset.ts

```diff
@@ -42,7 +47,7 @@ export class Dataset<T> implements AsyncIterable<T> {
    */
   map<U>(mapper: (_: T) => U | Promise<U>): Dataset<U> {
     const content = {
-      [Symbol.asyncIterator]: () => this.#content(),
+      [Symbol.asyncIterator]: () => this[Symbol.asyncIterator](),
```

Comment:
by using a method call, it will defer to the actual implementation of `Symbol.asyncIterator`. as it gets overriden in `CachingDataset`, we can now mix multiple implementation of the class and call the correct method. (note also that in `CachingDataset`, `this.#content` isn't valid because it's private to the parent class)